### PR TITLE
Backport [Bugs #18048]

### PR DESCRIPTION
**Pull request by nagachika**
Backport candidate for https://bugs.ruby-lang.org/issues/18048

### CODE REVIEW

**Comment by ioquatix** [comment]
Is it the same as https://github.com/ruby/ruby/pull/4686?

**Comment by ioquatix** [comment]
I reviewed both PRs, but they seem different. I'm not sure why.

File: thread.c [hunk 1]
```diff
@@ -734,38 +733,45 @@ thread_do_start_proc(rb_thread_t *th)
...
 
     switch (th->invoke_type) {
       case thread_invoke_type_proc:
+        result = thread_do_start_proc(th);
+        break;
+
       case thread_invoke_type_ractor_proc:
+        result = thread_do_start_proc(th);
+        rb_ractor_atexit(th->ec, result);
         thread_do_start_proc(th);
```
Comment (ioquatix) [inline thread.c:761]: You have `thread_do_start_proc` twice. I tried cherry-pick the given commits and the end result is different. This line is removed 287f2ba4f43d2557dd4f6168b31003b22a3b1ebd but not in this PR?

**Comment by ioquatix** [comment]
I did the following:

```
git cherry-pick 050a89543952a2c9e7c9bc938f4fdb538f6c9278
git cherry-pick 57eaa07ba6c1ee958c16d5c451e2dceb2208edf1
git cherry-pick edbe0e224c2594b7a7b055f0986cbfd690d754d5
git cherry-pick 2d4f29e77e883c29e35417799f8001b8046cde03
```

Minor mixup replacing `rb_fiber_scheduler_currrent` with `rb_scheduler_current` in `thread.c`. The end result is a little bit different from this PR.

**Comment by nagachika** [comment]
@ioquatix Thank you for your investigations! I will close this PR and merge https://github.com/ruby/ruby/pull/4686 instead.

**Comment by ioquatix** [comment]
Thanks for your time and assistance.

**Pull request closed without merging**